### DISCUSSION TITLE
raise error if conditions are not met

### DISCUSSION THREAD

**Discussion by ahmed711** [cue-lang/cue#1897] [Q&A] [answered]
Hello all, 
I am kind of new to cue and I was wondering about the right approach to raise errors at certain conditions 

I put up a scenario here on this playground
https://cuelang.org/play/?id=Z1x_8jWBXMB#cue@export@json

students is a list that consists of several students

```
students: {
[
    {
    name: "Adam"
    age: 15
    grade: 50
    status: "failed"
    },
    {
    name: "Akon"
    age: 15
    status: "passed"
    }
]
}

#student: {
    name: string
    age: number
    grade: number & ( 100 | 50 )
    status: string
    if (status =~ "passed"){
        grade: number & !=50
    }
    if (status =~ "failed"){
        grade: number & != 100
    }
}
students: [...#student]
```
This works fine for all cases except when the "grade" field is not there it doesn't generate an error that the field is missing, I want to generate an error if the grade field is not present

**Comment by ahmed711** [comment]
This is how I was able to workaround it but I hope there is a better way because the errors in this case are not intuitive

```
#student: {
    name: string
    age: number
    grade: number & ( 100 | 50 )
    status: string
    #passed: status == "passed"
    #failed: status =="failed"
    if (#passed){
        #passed: grade == 100
    }
    if (#failed"){
        #failed: grade == 50
    }
}
``` 

or less intuitive

```
#student: {
    name: string
    age: number
    grade: number & ( 100 | 50 )
    status: string
    #passed: status == "passed"
    if (#passed){
        #passed: grade == 100
    }
    if !(#passed){
        #passed: grade != 50
    }
}
```

**Comment by slewiskelly** [comment, marked as answer]
In your original example you've constrained `grade` to `100 | 50` but then you further constrain its value based on the value of `status`. 

`number & (100 | 50) & != 50`  can only be one value: `100`.

Removing these conditional constraints results in errors that you expect:

```cue
#student: {
	name:   string
	age:    int
	grade:  100 | 50
	status: "passed" | "failed"
}

students: [...#student] & [{
	name:   "Adam"
	age:    15
	grade:  50
	status: "failed"
}, {
	name:   "Akon"
	age:    15
	status: "passed"
}]
```

More generally, I would expect `status` to be constrained depending on the value of `grade`, as opposed to the other way around

```cue
#student: {
	name:   string
	age:    int
	grade:  100 | 50
	status: "passed" | "failed"

	if grade == 50 {
		status: "failed"
	}

	if grade == 100 {
		status: "passed"
	}
}

students: [...#student] & [{
	name:   "Adam"
	age:    15
	grade:  50
	status: "failed"
}, {
	name:   "Akon"
	age:    15
	grade:  100
}]
```

This obviously raises an error if `grade` has no value, but also constrains `status` if it has been input, but is not required.

**Reply by ahmed711** [reply]
@slewiskelly 
Many thanks for your reply 

I understand the constraint points but I want to be able to raise an error if the value is not there, the main use case is that I am doing this validation on some config files on git, and I want to have an error if the field is not on the file even if it is constrained already.

so in your example for instance if status are not there, I want to raise an error that it is not even if it is obviously passed as the grade is 100, and maybe that is a problem with our configuration files because if we know the value, then there is no point of forcing the user to enter it, but because the real example is a bit more complicated, we wanted to be able to raise an error if the field is not there

so when we ` cue vet schema config` we receive an error that the field is missing

**Reply by slewiskelly** [reply]
This is the cleanest way I can think of, given your requirements:

```cue
#student: {
	name:   string
	age:    int
	grade:  100 | 50
	status: "passed" | "failed"

	_valid: true &
		(grade == 100 && status == "passed") ||
		(grade == 50 && status == "failed")
}

students: [...#student] & [{
	name:   "Adam"
	age:    15
	grade:  50
	status: "failed"
}, {
	name:   "Akon"
	age:    15
	status: "passed"
}]

```

**Reply by ahmed711** [reply]
That is outstanding 
Thanks @slewiskelly 

I was looking for a way to throw errors and this `true & expression` was exactly what I was looking for 
Thanks again

**Reply by MAttia2022** [reply]
How do you throw and error in cue ( with a custom erorr message) or how do you change the default validation error message?

**Reply by mvdan** [reply]
@MAttia2022 see https://github.com/cue-lang/cue/issues/943.

**Reply by MAttia2022** [reply]
apparently, must is not defined as a builtin func in the version I use, Is there any other solution?

**Reply by mvdan** [reply]
That is an open proposal, not yet implemented. That's what you are after, so I don't think aything like it exists already.
